pkg/proxy/proxy.go

```diff
@@ -632,7 +812,10 @@ func replaceTokenInWebsocketRequest(req *http.Request, newToken string) {
 	req.Header.Set(ph, strings.Join(protocols, ","))
 }
 
-func validateWorkspaceRequest(requestedWorkspace, requestedNamespace string, workspaces []toolchainv1alpha1.Workspace) error {
+// validateWorkspaceRequest checks whether the requested workspace is in the list of workspaces the user has visibility on (retrieved via the spaceLister).
+// If `requestedWorkspace` is zero, this function looks for the home workspace (the one with `status.Type` set to `home`).
+// If `requestedNamespace` is NOT zero, this function checks if the namespace exists in the workspace.
```

Comment:
IIUIC, if we remove this check we could enable unwanted cross-workspace.

Consider the scenario in which Alice has access to workspaces `alice` - with namespace `alice-tenant`- and `bob` - with namespace `bob-tenant`. Both living in the same cluster.
If we remove the check, Alice could explicitly target the workspace `alice` and access the namespace `bob-tenant`. Something like `.../workspaces/alice/api/v1/namespaces/bob-tenant/pods`. Am I correct?

Reply:
> If we remove the check, Alice could explicitly target the workspace alice and access the namespace bob-tenant. Something like `.../workspaces/alice/api/v1/namespaces/bob-tenant/pods`. Am I correct?

Not really. Our proxy will redirect the requests indeed to the same cluster: `/api/v1/namespaces/bob-tenant/pods` as Alice. But since Alice does not have permissions to access that namespace, kube will reject it.

It basically will still fail but on the kube side. Not on the proxy side.

Reply:
ok, but if Alice has access to `bob-tenant` namespace because Bob gave her access to his workspace, the request will not fail. Right?

If this is correct, Alice could set a workspace and perform cross-workspace requests by simply using a namespace from another workspace. The only requirements are:
1. workspaces are hosted on the same cluster
2. Alice has access to both workspaces

So Alice can have a kubeconfig with server url set to `https://cluster-domain/workspaces/alice-workspace` and target namespaces from different workspace. To me this is a problem, because as a user I expect my requests to be confined to the workspace I set in the url.

Reply:
This is correct but I still don't see a problem with that. It's not like Alice can get hack the proxy and get access to a namespace she doesn't have access to. She just could create a weird workspace/namespace context and use `/workspaces/alice/api/v1/namespaces/bob-tenant/pods` instead of more appropriate `/workspaces/bob/api/v1/namespaces/bob-tenant/pods`.

This is exactly why we introduce it in the first place. There was a bug in HAC/UI with the workspace switching. The workspace switcher would change the workspace but wouldn't change the namespace. So we hopped that setting such a limit would make it easy to spot bugs/mistakes on the client side (like in HAC): hey, you are using a weird context, make sure you really knows what namespace in what workspace you want to access otherwise you could be scratching your head wondering why it gives you weird result.
So it is not a security problem.

However it has significant side effects. We faced multiple cases when users try to access some namespaces which do not belong to any workspace at all! Some global view-only namespaces which should be available to all authenticated users. Some operators use such approach.
With this limitation in place it's impossible to access such non-workspace namespaces through the proxy even if they are totally accessible using the api server directly.

Reply:
Here is the followup PR: https://github.com/codeready-toolchain/registration-service/pull/471